util/provenance/buildinfo.go

```diff
@@ -0,0 +1,164 @@
+package provenance
+
+import (
+	"encoding/hex"
+	"strings"
+
+	distreference "github.com/docker/distribution/reference"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+	binfotypes "github.com/moby/buildkit/util/buildinfo/types"
+	digest "github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+var BuildKitBuildType = "https://mobyproject.org/buildkit@v1"
+
+type ProvenancePredicate struct {
+	slsa.ProvenancePredicate
+	Metadata *ProvenanceMetadata                `json:"metadata,omitempty"`
+	Source   *Source                            `json:"buildSource,omitempty"`
+	Layers   map[string][][]ocispecs.Descriptor `json:"buildLayers,omitempty"`
+}
+
+type ProvenanceMetadata struct {
+	slsa.ProvenanceMetadata
+	VCS map[string]string `json:"vcs,omitempty"`
+}
+
+func convertMaterial(s binfotypes.Source) (*slsa.ProvenanceMaterial, error) {
+	// https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst
```

Comment:
Can we use `RefToPURL` here?

Reply:
I've rewritten this part in follow-up branch https://github.com/moby/buildkit/compare/master...tonistiigi:buildkit:attestations-provenance-buildinfo

Reply:
SGTM, this is fine for now then.